libcrux-ml-dsa/src/simd/portable/encoding/commitment.rs

```diff
@@ -1,40 +1,161 @@
+            hax_lib::loop_invariant!(|i: usize| {fstar!(r"
+            Seq.length $serialized == 4 /\ (
+              let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 4 in
+              let out = bit_vec_of_int_t_array #U8 #(mk_usize 4) $serialized 8 in
+              forall (n: nat {n < v i * 8}). out n == inp n
+            )")});
+
+            serialized[i] = encode_4(coefficients);
 
-        6 => {
-            // The commitment has coefficients in [0,43] => each coefficient occupies
-            // 6 bits.
-            cloop! {
-                for (i, coefficients) in simd_unit.values.chunks_exact(4).enumerate() {
-                    let coefficient0 = coefficients[0] as u8;
-                    let coefficient1 = coefficients[1] as u8;
-                    let coefficient2 = coefficients[2] as u8;
-                    let coefficient3 = coefficients[3] as u8;
-
-                    serialized[3 * i] = (coefficient1 << 6) | coefficient0;
-                    serialized[3 * i + 1] = (coefficient2 << 4) | coefficient1 >> 2;
-                    serialized[3 * i + 2] = (coefficient3 << 2) | coefficient2 >> 4;
-                }
-            }
-            ()
+            hax_lib::fstar!(r"
+                let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 4 in
+                let out = bit_vec_of_int_t_array #U8  #(mk_usize 4) $serialized 8 in
+                introduce forall (n:nat{n < 8}). inp (v i * 8 + n) == out (v i * 8 + n)
+                with
+                  (calc (==) {
+                          inp (v i * 8 + n);
+                    == {} get_bit (Seq.index ${simd_unit.values} ((v i * 8 + n) / 4)) (sz ((v i * 8 + n) % 4));
+                    == { Math.Lemmas.division_addition_lemma n 4 (v i * 2) }
+                          get_bit (Seq.index ${simd_unit.values} (v i * 2 + n / 4)) (sz (n % 4));
+                    == {} get_bit (Seq.index $coefficients (n / 4)) (sz (n % 4));
+                    == {} bit_vec_of_int_t_array #I32 #(mk_usize 2) $coefficients 4 n;
+                    == {} out (v i * 8 + n);
+                  })
+            ");
         }
+    }
+    ()
+}
+
+#[inline(always)]
+#[hax_lib::requires(fstar!(r"Seq.length ${serialized} == 6 /\ (forall i. bounded (Seq.index ${simd_unit.values} i) 6)"))]
+#[hax_lib::ensures(|out| {
+    let serialized_future = future(serialized);
+    fstar!(r"
+Seq.length ${serialized_future} == Seq.length ${serialized} /\
+(let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 6 in
+ let out = bit_vec_of_int_t_array #U8  #(mk_usize 6) ${serialized_future} 8 in
+ forall (i: nat {i < 8 * 6}). inp i == out i)
+")
+})]
+pub fn serialize_6(simd_unit: &Coefficients, serialized: &mut [u8]) {
+    // The commitment has coefficients in [0,43] => each coefficient occupies
+    // 6 bits.
+    for (i, coefficients) in simd_unit.values.chunks_exact(4).enumerate() {
```

Comment:
This needs a `cloop` for eurydice extraction.

Reply:
done